Oracle case study

loiclefevre opened the pull request:
**What kind of change does this PR introduce?**

Adds the Oracle case study to the blog.

**Issue Number:**

N/A?

**Summary**

This PR adds the Oracle case study which presents numerous supported JSON Schema capabilities along code examples.

**Does this PR introduce a breaking change?**

No

jviotti commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1123 @@
+tags:
+  - database
+  - relational
+type: Case Study
```
It might be just me, but I think the article goes beyond the typical case study (i.e. here is how company X uses JSON Schema) into actually teaching something to the reader. From that point of view, maybe better to post it on the "Engineering" category?

jviotti replied:
On the other side, I think Oracle's case study is indeed very technical, so maybe that's fine. Not sure. @benjagm what do you think?

jviotti commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1123 @@
+---
+title: "JSON Schema and Relational Databases: Bridging the Gap!"
```
I think your title should directly mention Oracle!

gregsdennis replied:
Agreed. Look at the other case studies we have on our site as examples.

loiclefevre replied:
New title: JSON Schema and Relational Databases: Bridging the Gap with Oracle!

gregsdennis replied:
Maybe I'm being picky, but that sounds to me like this is about how to use Oracle rather than how Oracle is using JSON Schema to solve problems.

gregsdennis replied:
Going over the post, I see it is a bit different from our other case studies.

I think something more focused on the company, like "How Oracle is Bridging the Gap Between JSON Schema and Relational Databases".  Bringing the company to the front highlights it more, then you find out what they're doing.

jviotti commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1123 @@
+{
+  "type" : "object",
+  "o:length" : 1,
+  "properties" :
```
Minor thing, but can you avoid opening brackets, etc on separate lines in the examples? Just for consistence with other blog posts

loiclefevre replied:
I see what you mean, this is just because this is the output/result of the command above: `json_dataguide()` function. I'll fix this.

jviotti commented:
I left a few comments, but overall I really like this blog post. Very interesting stuff on directly using JSON Schema for something cool that I think our readers will enjoy.

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+excerpt: ""
+---
+
+As modern multi-model databases increasingly support JSON, it's time to explore what role [JSON schema](https://json-schema.org/) will play. In this post, we'll dive into the newly developed ["Database Vocabulary"](https://github.com/json-schema-org/vocab-database/blob/main/database.md), a proposal to the official JSON schema specification developed by Oracle (with inputs from the MySQL and Postg ...
```
> a proposal to the official JSON Schema specification developed by Oracle 

Sounds like Oracle is claiming they developed JSON Schema. 

Also, this is a proposed _extension_, not an amendment to the spec.

loiclefevre replied:
I'll add your suggestion and a comma to make it more clear 😅

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
````diff
@@ -0,0 +1,1112 @@
+select data from blog_posts;
+```
+
+|DATA|
+|-|
+|{"title":"New Blog Post","content": "This is the content of the blog post...","publishedDate":"2023-08-25T15:00:00Z","author":{"username":"authoruser","email":"author@example.com"},"tags":["Technology","Programming"]}|
+|{"garbageDocument":true}|
````
I don't think this is going to render well. A long string in a table isn't going to be great.  Is there another way this can be shown?

loiclefevre replied:
I can definitely reformat the JSON doc inside the table.

loiclefevre replied:
Would render like this be okay?

![image](https://github.com/user-attachments/assets/297926b9-107d-4ea3-80d1-9fe96fbb4b16)

gregsdennis replied:
This looks good to me. Again, let's have the web folks verify.

loiclefevre replied:
In fact, I've tested it using the preview button [View deployment](https://212d361f.website-2v2.pages.dev/)

loiclefevre replied:
Hence I've modified it so that now it renders as: 
![image](https://github.com/user-attachments/assets/53ddae55-ed47-42b7-b16d-440d71d9c0a0)

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
````diff
@@ -0,0 +1,1112 @@
+from blog_posts;
+```
+
+|REPORT|
+|-|
+|{"valid": true,"errors": []}|
+|{"valid": false,	"errors": [	{	"schemaPath": "\$", "instancePath": "\$", "code": "JZN-00501", "error": "JSON schema validation failed" }, { "schemaPath": "\$.required", "instancePath": "\$", "code": "JZN-00515", "error": "required properties not found: 'title', 'content', 'author'" }	] }|
````
Again, I think the format should be reconsidered.

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
````diff
@@ -0,0 +1,1112 @@
+select dbms_json_schema.describe( 'BLOG_POSTS' ) as json_schema;
+```
+
+|JSON_SCHEMA|
+|-|
+|{"title": "BLOG_POSTS", "dbObject": "APIDAYS.BLOG_POSTS", "type": "object", "dbObjectType": "table", "properties": { "DATA": {} } }|
````
Format

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+
+With all this in place, our React frontend can now create the following form:
+
+<figure class="mt-10">
+  <img class="rounded-lg" src="/img/posts/2025/oracle-case-study/form.webp"/>
+  <figcaption class="mt-2 mb-10 text-sm text-center text-gray-500">React frontend with input form generated from an annotated Oracle Database 23ai JSON schema.</figcaption>
+</figure>
```
I think we have special markdown for components like this. Please check the blog post style guide.

loiclefevre replied:
but maybe this is just this markdown component you are mentioning: `![...](...)`

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+
+Below, you can see a quick overview of a demo available in this [GitHub repository](https://github.com/loiclefevre/apidays-paris-2024):
+
+<figure class="mt-10">
+  <img src="/img/posts/2025/oracle-case-study/json-schema-form.webp"/>
+  <figcaption class="mt-2 mb-10 text-sm text-center text-gray-500">React frontend form created from the json-schema-form library.</figcaption>
+</figure>
```
Check for special markdown

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+  <figcaption class="mt-2 mb-10 text-sm text-center text-gray-500">React frontend with input form generated from an annotated Oracle Database 23ai JSON schema.</figcaption>
+</figure>
+
+Remark: interestingly, whenever you change the schema annotation in the database, it is immediately reflected inside your browser once you refreshed it. You can try with:
```
We definitely have special components for these kinds of sidebar notes.

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+
+Running the 2 queries above respectively returns the data in JSON and relational formats:
+
+|DATA|
+|-|
+|{"_id": "Cake mould", "PRICE": 9.99, "QUANTITY": 15, "_metadata": { ... } }|
+|{"_id": "Wooden spatula", "PRICE": 4.99, "QUANTITY": 42, "_metadata": { ... } }|
+|{"_id": "Other nice product", "PRICE": 5, "QUANTITY": 10, "_metadata": { ... } }|
```
Reconsider the format

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+
+Once a check constraint is marked as `PRECHECK`, you have the choice whether or not to disable the check constraint on the table as the retrieved JSON schema with `dbms_json_schema.describe()` will contain the check constraints as well.
+
+**WARNING**: we do *NOT* advise to disable check constraints as it would allow inserting bad data into the relational tables directly. The remark about `PRECHECK` constraints is here to provide as much information as possible.
```
Sidebar component

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
````diff
@@ -0,0 +1,1112 @@
+select * from orders;
+```
+
+|J|J$address|J$firstName|J$vat|
+|-|-|-|-|
+|{'firstName':'Bob', 'address' : 'Paris'}|Paris|Bob|null|
+|{'firstName':'Bob', 'address' : 'Paris', 'vat': false}|Paris|Bob|false|
````
Reconsider format

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
````diff
@@ -0,0 +1,1112 @@
+select * from orders;
+```
+
+|J|J$address|J$firstName|J$vat|J$tableEvolve|
+|-|-|-|-|-|
+|{'firstName':'Bob', 'address' : 'Paris'}|Paris|Bob|null|null|
+|{'firstName':'Bob', 'address' : 'Paris', 'vat': false}|Paris|Bob|false|null|
+|{'firstName':'Bob', 'address' : 'Paris', 'vat': false, 'tableEvolve': true}|Paris|Bob|false|true|
````
Reconsider format

gregsdennis commented on pages/blog/posts/oracle-case-study.md:
```diff
@@ -0,0 +1,1112 @@
+
+We have shown lots of features inside the Oracle Database 23ai which provide powerful capabilities to have JSON data coexist with relational data, and JSON schema clearly strengthens this even more. But this is only the beginning and as you discover more and more features that work the same way regardless of the data model, or that allow going back and forth from one model to another, you'll unde ...
+
+<figure class="mt-10">
+  <img class="rounded-lg" src="/img/posts/2025/oracle-case-study/converged_database.webp"/>
+  <figcaption class="mt-2 mb-10 text-sm text-center text-gray-500">Oracle Database 23ai is a converged database now supporting JSON schema.</figcaption>
+</figure>
```
Use component

loiclefevre replied:
Please provide an example

gregsdennis reviewed (changes requested):
I think the content is fine, though I don't think we've had a case study quite like this. Normally they're more, "how JSON Schema helped us solve a problem."  In this case, it's more, "how we're using JSON Schema to provide more features" which is fine. 

There are a number of stylistic issues, and I think we should verify the preview and have one or more of the web admins have a look.  There should be a style guide somewhere, but I'm on mobile and don't have immediate access to it.  You can also look at some of the other posts for examples.

loiclefevre commented:
This preview thing is terrific!

cbum-dev commented:
I think some more spacing should be their to easily read the outputs.
Can I help here?

gregsdennis commented:
@cbum-dev There are definitely some style things that need to be addressed in general, but that needs to happen outside of this PR since it's work on the site itself.

For this PR, we need to ensure the content matches our style guide:

- using components that we've defined
- language is clear
- etc.

@loiclefevre for reference, here are our [blog guidelines](https://github.com/json-schema-org/community/blob/main/docs/blog-guidelines.md).

@benjagm I recall asking for a page like https://chirpy.cotes.page/posts/write-a-new-post/.  Did we ever do that?  I can't find that issue now.

loiclefevre commented:
@gregsdennis, I miss some examples indeed with pictures and Danger, Tip, InfoBox.

I've tried them and here is my feedback:
- image defined at the top doesn't support full transparency when inside the blog post (white background forced)
- All boxes make their text displayed on the left completely inside the navigation menu
- Labels rendering is not great
- InfoBox or Tip background color is the same as for inlined code making a weird effect where we do not distinguish inlined code anymore

Hence I've tried to work around those as best as I could.

gregsdennis commented:
I think for the purposes of this PR, we want to use the elements we've defined.  If there are styling issues with those (and it seems there are), then the web team needs to address those separately.

What I don't want is to have to update your blog post. I'd rather have the blog post set up correctly, then when the site is fixed, your post fixes itself.

I do recognize that there is an element of branding and Oracle representation, but it's on our site, so we need to take that responsibility.

gregsdennis commented:
Okay, my suggestion is to revert that last commit, but then open a separate PR with its content (the one with all the problems).  We'll merge for now and start working on the issues. Once we've fixed the issues, then we have the second PR that we can just merge.

cbum-dev commented:
> @gregsdennis, @cbum-dev, I've pushed a new version where you can find the different issues mentioned:
> 
>     * Infobox with text replicated in the left margin:
>       ![image](https://private-user-images.githubusercontent.com/21245376/410383103-355a1a9b-5867-41f1-9037-8a0408d0582f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODMxMDMtMzU1YTFhOWItNTg2Ny00MWYxLTkwMzctOGEwNDA4ZDA1ODJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzZGYyN2RkY2E2MmRlZTFhNGViM2QzN2M5NGEzNGFjMmU5MTkzMGFjMjA0YzBlNDY4NzBkNWZhNDdiNWEwM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SKOFQM2DTNd6Z09BmMN-c0ud_7zR72sCpwfCSjoDw_Y)
> 
>     * Infobox not supporting markdown block for source code:
>       ![image](https://private-user-images.githubusercontent.com/21245376/410383353-d5a6e088-3161-4875-aba3-13648cae51f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODMzNTMtZDVhNmUwODgtMzE2MS00ODc1LWFiYTMtMTM2NDhjYWU1MWY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxMmFiNGVlYzRjMGUwMjFlZGY5NDRiM2Q5NTVlODgxYzllNDVmMmU3ZDlmZTFlMTMyNTNkNGFhNmE4OWE2ZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Q7rmkEaQGn8-sQtr81RpM4bdIjqLWrJYEmxgFY_XTPo)
> 
>     * Infobox background color similar to inlined markdown code block:
>       
>       * the _metadata word should appear with a blue background color that highlights it's some specific keyword, but the background colors are similar, hence the highlighting fails
>         ![image](https://private-user-images.githubusercontent.com/21245376/410385991-df219536-57f6-4d5a-8dae-7e0932c4a141.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODU5OTEtZGYyMTk1MzYtNTdmNi00ZDVhLThkYWUtN2UwOTMyYzRhMTQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlMzdiZGU0MDY5YjBhYzBiNWM3YmM3YmU4NTI5Yjg0MTVlOWFiZDNhNjYwMmE3YmE3MDUwMjRjMGY0NTJmNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ff8nqKsikwjJskEqs6zB9iNULsMeYk5M6bZ2BFok6Qg)
> 
>     * Fully transparent cover picture ends up with a white background color:
>       ![image](https://private-user-images.githubusercontent.com/21245376/410384816-c5308aa7-d0e3-4543-9a46-5e6e04f93960.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODQ4MTYtYzUzMDhhYTctZDBlMy00NTQzLTlhNDYtNWU2ZTA0ZjkzOTYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM5NmJlNmU0ZDYwN2RhNTgyZDY3ZDkxN2IyZWQwMzVlMzIxNWE3ODRjZTljZmM1MjY1ZTVhNTZiMTRkYzVjOGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ph6rNnpBxteS6UBfAvS2qseU4LTAVo7Xe2yjjHrd9Fs)
> 
>     * Hard to read the white text over the cover picture:
>       ![image](https://private-user-images.githubusercontent.com/21245376/410385529-3674fc85-0ba2-47a8-b8be-8985f30693c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODU1MjktMzY3NGZjODUtMGJhMi00N2E4LWI4YmUtODk4NWYzMDY5M2MxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwOTFmNDQ0Y2RhN2MzMTc0YzhmZjZlZDc5YzdkMzhkYzQ1ZGI0OTYyNTc1ZmFiMzRlZTYyMmRjZmU5YjMwOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iVNdDctoSWobfaX7VZAhEEFMi-ovV1V8kW-UgjoV0bQ)
> 
>     * Danger with text replicated in the left margin:
>       ![image](https://private-user-images.githubusercontent.com/21245376/410386768-36d14d5a-55f8-4ccd-a40a-d89a49b390e7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODY3NjgtMzZkMTRkNWEtNTVmOC00Y2NkLWE0MGEtZDg5YTQ5YjM5MGU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVmOThlNGYwMDFlMTI3OTE5OWNmZWJlODU0MjJmNWMzZWVlOThhZmNmYjhlNjY1ZjU0ZDY1MTM4NGIyM2Y2MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uLVz1bYNMVc5pA7wHc-cYbcYtimBQ1aXZ7XlxvvNCT0)
> 
>     * Markdown bolded text in Danger component not rendered properly:
>       ![image](https://private-user-images.githubusercontent.com/21245376/410387369-10f343c9-7f93-4c00-ac13-8cea76a2046b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDE1NjcsIm5iZiI6MTczODg0MTI2NywicGF0aCI6Ii8yMTI0NTM3Ni80MTAzODczNjktMTBmMzQzYzktN2Y5My00YzAwLWFjMTMtOGNlYTc2YTIwNDZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDExMjc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiODU5OGZmNTk5M2I2ZDNjOWU3MTAyNDNkOWFhZjcyMjk3N2E5ZGZlMTg4NmE1MjVjZTM4MDJiNDI4MDcyYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.smy85ONAOK-5clG085u09aiiq5yQTokbahiQ-H-siaA)
> 
> 
> I hope these get fixed :)

The readability issue was already opened by me https://github.com/json-schema-org/website/issues/1277 but waiting for the team's approval .

loiclefevre commented:
> Okay, my suggestion is to revert that last commit, but then open a separate PR with its content (the one with all the problems). We'll merge for now and start working on the issues. Once we've fixed the issues, then we have the second PR that we can just merge.

Hi @gregsdennis, I've reverted to the last commit! I'm working on the new PR now...
